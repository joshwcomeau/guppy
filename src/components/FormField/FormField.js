// @flow
import React, { PureComponent } from 'react';
import styled from 'styled-components';

import { RAW_COLORS, COLORS } from '../../constants';

import Label from '../Label';

type Props = {
  label: string,
  useLabelTag?: boolean,
  isFocused?: boolean,
  hasError?: boolean,
  children: React$Node,
  spacing: number,
};

class FormField extends PureComponent<Props> {
  static defaultProps = {
    spacing: 30,
  };

  render() {
    const {
      label,
      useLabelTag,
      isFocused,
      hasError,
      children,
      spacing,
    } = this.props;

    const Wrapper = useLabelTag ? WrapperLabel : WrapperDiv;

    return (
      <Wrapper spacing={spacing}>
        <LabelText isFocused={isFocused} hasError={hasError}>
          {label}
        </LabelText>
        <Children>{children}</Children>
      </Wrapper>
    );
  }
}

const getTextColor = (props: Props) => {
  if (props.hasError) {
    return COLORS.lightError;
  } else if (props.isFocused) {
    return RAW_COLORS.purple[700];
  } else {
    return RAW_COLORS.gray[500];
  }
};

const WrapperLabel = styled.label`
  display: block;
  margin-bottom: ${props => props.spacing}px;
`;

const WrapperDiv = styled.div`
  margin-bottom: ${props => props.spacing}px;
`;

const LabelText = styled(Label)`
  flex-basis: ${props => props.width}px;
  color: ${getTextColor};
  padding: 0px 5px;
`;

const Children = styled.div`
  flex: 1;
  padding: 0px 5px 5px;
`;

export default FormField;
