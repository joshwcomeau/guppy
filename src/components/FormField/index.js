// @flow
export { default } from './FormField';
