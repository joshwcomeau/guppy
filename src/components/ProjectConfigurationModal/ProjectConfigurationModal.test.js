/* eslint-disable flowtype/require-valid-file-annotation */
import React from 'react';
import { shallow } from 'enzyme';
import {
  ProjectConfigurationModal,
  initialState,
} from './ProjectConfigurationModal';
import FormField from '../FormField';

describe('ProjectConfigurationModal component', () => {
  let wrapper;
  let instance;
  let mockActions;

  const project = {
    id: 'a-project',
    name: 'A project',
    projectIcon: 'icon',
  };

  const shallowRender = (installActive = false) => {
    mockActions = {
      hideModal: jest.fn(),
      saveProjectSettings: jest.fn(),
    };
    return shallow(
      <ProjectConfigurationModal
        isVisible={true}
        project={project}
        dependenciesChangingForProject={installActive}
        {...mockActions}
      />
    );
  };

  describe('Rendering', () => {
    it('should render', () => {
      wrapper = shallowRender(false);
      expect(wrapper).toMatchSnapshot();
    });

    it('should render with active dependency installation (save disabled)', () => {
      wrapper = shallowRender(true);
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('Component logic', () => {
    const newProject = {
      name: 'new-project',
      icon: 'another-icon',
    };

    beforeEach(() => {
      wrapper = shallowRender();
      instance = wrapper.instance();
    });

    it('should update local form state for new project', () => {
      instance.componentWillReceiveProps({
        project: newProject,
      });
      // Note: If there would be more options we could use a project key on state
      //       so we can easily compare the object here. Or we could use a loop to check each key -
      //       for now it'S OK to repeat the assertions because there are just two keys.
      expect(instance.state.newName).toEqual(newProject.name);
      expect(instance.state.projectIcon).toEqual(newProject.icon);
    });

    it('should not update local form state if project undefined', () => {
      // project is always defined - this is needed for flow type check
      instance.componentWillReceiveProps({ project: null });
      expect(instance.state).toBe(initialState);
    });

    it('should save new settings', () => {
      instance.componentWillReceiveProps({
        project: newProject,
      });
      // Note: Simulate click on save button wasn't working
      //       that's why I'm submitting the form
      wrapper.find('form').simulate('submit', {
        preventDefault: jest.fn(),
      });

      expect(mockActions.saveProjectSettings).toHaveBeenCalledWith(
        instance.state.newName,
        instance.state.projectIcon,
        project
      );
    });

    describe('Form handling', () => {
      it('should save settings on enter key', () => {
        instance.handleKeyPress({
          preventDefault: jest.fn(),
          key: 'Enter',
        });
        expect(mockActions.saveProjectSettings).toHaveBeenCalled();
      });

      it('should not save on other keypress', () => {
        instance.handleKeyPress({
          preventDefault: jest.fn(),
          key: 'a',
        });
        expect(mockActions.saveProjectSettings).not.toHaveBeenCalled();
      });

      it('should updateProject icon in local state', () => {
        instance.updateProjectIcon('new-icon', {
          preventDefault: jest.fn(),
        });
        expect(instance.state.projectIcon).toEqual('new-icon');
      });

      it('should set a form field to active in local state', () => {
        instance.setActive('projectName');
        expect(instance.state.activeField).toEqual('projectName');
      });

      it('should set activeField to projectName on text input focus', () => {
        instance.setActive('projectIcon');
        expect(instance.state.activeField).toEqual('projectIcon');
        const input = wrapper
          .find(FormField)
          .find('[label="Project name"]')
          .children();
        input.prop('onFocus')();

        expect(instance.state.activeField).toEqual('projectName');
      });

      it('should change newName on input change', () => {
        instance.changeProjectName({
          currentTarget: {
            value: 'New name',
          },
        });

        expect(instance.state.newName).toEqual('New name');
      });
    });
  });
});
