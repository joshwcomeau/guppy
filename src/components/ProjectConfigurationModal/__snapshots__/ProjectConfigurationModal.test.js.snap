// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectConfigurationModal component Rendering should render 1`] = `
<Modal
  isVisible={true}
  onDismiss={[MockFunction]}
  width={750}
>
  <ModalHeader
    theme="standard"
    title="Project settings"
  />
  <styled.section>
    <form
      onSubmit={[Function]}
    >
      <FormField
        focusOnClick={false}
        label="Project name"
        spacing={30}
      >
        <TextInput
          autoFocus={true}
          isFocused={true}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyPress={[Function]}
          value=""
        />
      </FormField>
      <styled.div
        size={10}
      />
      <FormField
        focusOnClick={false}
        isFocused={false}
        label="Project Icon"
        spacing={30}
      >
        <ProjectIconSelection
          limitTo={21}
          onSelectIcon={[Function]}
          selectedIcon=""
        />
      </FormField>
      <styled.div>
        <FillButton
          colors={
            Array [
              "#00C853",
              "#69db0d",
            ]
          }
          disabled={false}
          size="large"
          textColor="#FFF"
        >
          Save Project
        </FillButton>
      </styled.div>
    </form>
  </styled.section>
</Modal>
`;

exports[`ProjectConfigurationModal component Rendering should render with active dependency installation (save disabled) 1`] = `
<Modal
  isVisible={true}
  onDismiss={[MockFunction]}
  width={750}
>
  <ModalHeader
    theme="standard"
    title="Project settings"
  />
  <styled.section>
    <form
      onSubmit={[Function]}
    >
      <FormField
        focusOnClick={false}
        label="Project name"
        spacing={30}
      >
        <TextInput
          autoFocus={true}
          isFocused={true}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyPress={[Function]}
          value=""
        />
      </FormField>
      <styled.div
        size={10}
      />
      <FormField
        focusOnClick={false}
        isFocused={false}
        label="Project Icon"
        spacing={30}
      >
        <ProjectIconSelection
          limitTo={21}
          onSelectIcon={[Function]}
          selectedIcon=""
        />
      </FormField>
      <styled.div>
        <FillButton
          colors={
            Array [
              "#00C853",
              "#69db0d",
            ]
          }
          disabled={true}
          size="large"
          textColor="#FFF"
        >
          Save Project
        </FillButton>
        <styled.div>
          Waiting for pending tasks to finishâ€¦
        </styled.div>
      </styled.div>
    </form>
  </styled.section>
</Modal>
`;
